### PR TITLE
Blog

### PR DESCRIPTION
**Pull request by davish**
This PR will add the blog to the new site.

Ghost Authors are not displayed on the site; instead, they are mapped to a Labs Member instead. This might become an issue if down the line we want to have guest posts from people outside of Labs. But we could just make a team called "Guests" and not display them on the team page.

- [ ] See if there's a way to link `GhostPostAuthor` type to `Member`... doesn't seem to be though.
- [x] Factor out / generalize components
- [x] Style tag pages
- [x] Display tags on post page and in post list
- [x] Pagination
- [ ] Make sure SEO metadata from ghost is getting where it needs to go

### Blog Index
Screen Shot 2020-01-25 at 5 29 58 PM

### Blog Index (bottom, showing pagination)
image

### Post (top)
Screen Shot 2020-01-25 at 5 30 10 PM

### Post (bottom)
Screen Shot 2020-01-25 at 5 30 51 PM

### Member page (bottom)
Screen Shot 2020-01-25 at 5 31 13 PM

### Tag detail page
Screen Shot 2020-01-25 at 5 31 32 PM

### CODE REVIEW

File: gatsby-config.js [hunk 1]
```diff
@@ -39,6 +39,15 @@ module.exports = {
         },
       },
     },
+    {
+      resolve: `gatsby-source-ghost`,
+      options: {
+        apiUrl: `https://ghost.pennlabs.org`,
+        contentApiKey: process.env.GHOST_API_KEY,
+        version: `v3`, // Ghost API version, optional, defaults to "v3".
+        // Pass in "v2" if your Ghost install is not on 3.0 yet!!!
```
Comment (cacabo) [inline gatsby-config.js]:
```suggestion
```

File: gatsby-node.js [hunk 2]
```diff
@@ -115,13 +117,130 @@ exports.sourceNodes = async ({ actions }) => {
...
-// exports.onCreateNode = ({ actions, node }) => {
-//   const { type } = node.internal
-//   if (type !== `Member`) return
-//   const { bio } = node
-//   const bioAsHtml = markdownProcessor.process(bio)
-//   node.bio = bioAsHtml
-// }
+exports.onCreateNode = ({ node, actions }) => {
+  const { type } = node.internal
+  if (type === 'GhostAuthor') {
```
Comment (cacabo) [inline gatsby-node.js]: Might be good to make all node types constants at the top of the file

File: gatsby-node.js [hunk 3]
```diff
@@ -115,13 +117,130 @@ exports.sourceNodes = async ({ actions }) => {
...
+    if (node.slug !== 'data-schema-author') {
+      actions.createNodeField({
```
Comment (cacabo) [inline gatsby-node.js]: To be consistent w/ other functions either change all the other ones to use `actions.doSomething` or change this to be

```javascript
const { doSomething } = actions
...
doSomething()
```

File: gatsby-node.js [hunk 4]
```diff
@@ -115,13 +117,130 @@ exports.sourceNodes = async ({ actions }) => {
...
+        node,
+        name: `member___NODE`,
```
Comment (cacabo) [inline gatsby-node.js:138]: Can you include a comment with the webpages which document this approach? other ppl will have not much of an idea of what's going on when they look over dis code

File: gatsby-node.js [hunk 5]
```diff
@@ -115,13 +117,130 @@ exports.sourceNodes = async ({ actions }) => {
...
+        value: `Labs__Member__${node.slug}`,
+      })
+    }
+  }
+}
+
+const createTagPages = (tags, createPage) => {
+  const tagsTemplate = path.resolve(`./src/templates/tag.tsx`)
```
Comment (cacabo) [inline gatsby-node.js]: I think the other template is capitalized, so would recommend changing this to `...Tag.tsx`

File: gatsby-node.js [hunk 6]
```diff
@@ -115,13 +117,130 @@ exports.sourceNodes = async ({ actions }) => {
...
+      const nextPageNumber =
+        currentPage + 1 > numberOfPages ? null : currentPage + 1
+      const previousPagePath = prevPageNumber
+        ? prevPageNumber === 1
+          ? node.url
+          : `${node.url}page/${prevPageNumber}/`
+        : null
+      const nextPagePath = nextPageNumber
+        ? `${node.url}page/${nextPageNumber}/`
+        : null
```
Comment (cacabo) [inline gatsby-node.js]: Might be good to put this in some helper functions bc the ternaries all next to each other are kinda hard to parse

File: gatsby-node.js [hunk 7]
```diff
@@ -115,13 +117,130 @@ exports.sourceNodes = async ({ actions }) => {
...
+      createPage({
+        path: i === 0 ? node.url : `${node.url}page/${i + 1}/`,
+        component: authorTemplate,
+        context: {
+          // Data passed to context is available
+          // in page queries as GraphQL variables.
+          slug: node.slug,
+          limit: postsPerPage,
+          skip: i * postsPerPage,
+          numberOfPages: numberOfPages,
```
Comment (cacabo) [inline gatsby-node.js]:
```suggestion
          numberOfPages,
```

File: gatsby-node.js [hunk 8]
```diff
@@ -115,13 +117,130 @@ exports.sourceNodes = async ({ actions }) => {
...
+          humanPageNumber: currentPage,
+          prevPageNumber: prevPageNumber,
+          nextPageNumber: nextPageNumber,
+          previousPagePath: previousPagePath,
+          nextPagePath: nextPagePath,
```
Comment (cacabo) [inline gatsby-node.js]: Do the same for all of these (no need to repeat the same string--take advantage of js syntactic sugar)

File: gatsby-node.js [hunk 9]
```diff
@@ -115,13 +117,130 @@ exports.sourceNodes = async ({ actions }) => {
...
+        },
+      })
+    })
+  })
+}
+
+const createAuthorPages = authors => {
```
Comment (cacabo) [inline gatsby-node.js]: Also do we need author pages or are member pages enough? What's the difference?

File: gatsby-node.js (hunk 5 above)
Comment (cacabo) [reply gatsby-node.js]: Actually you can see what I did at the other one, declared the template at the top of the file and then used it down here. Imo that way of doing it is a little cleaner / decouples operating on the file system

File: gatsby-node.js [hunk 10]
```diff
@@ -137,20 +256,88 @@ exports.createPages = async ({ graphql, actions }) => {
...
+
+  // Check for any errors
+  if (result.errors) {
+    throw new Error(result.errors)
+  }
+
+  // Extract query results
+  const tags = result.data.allGhostTag.edges
+  const authors = result.data.allGhostAuthor.edges
+  const posts = result.data.allGhostPost.edges
```
Comment (cacabo) [inline gatsby-node.js]:
```js
const {
  allGhostTag: { edges: tags },
  allGhostAuthor: { edges: authors },
  allGhostPost: { edges: posts },
} = result.data
```

Comment (cacabo) [reply gatsby-node.js]: I'm a fan of destructuring here but do whatever you prefer

File: src/templates/blog.tsx [hunk 11]
```diff
@@ -0,0 +1,47 @@
+import React from 'react'
+
```
Comment (cacabo) [inline src/templates/blog.tsx]:
```suggestion
```

File: src/templates/blog.tsx [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+import { graphql, Link } from 'gatsby'
+
+
```
Comment (cacabo) [inline src/templates/blog.tsx]:
```suggestion
```

File: src/templates/blog.tsx [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+import Layout from '../components/Layout'
+import SEO from '../components/SEO'
+import { H1, Section, Container } from '../shared'
+
+const BlogPage = ({data}): React.ReactElement => {
+  const posts = data.allGhostPost.edges
```
Comment (cacabo) [inline src/templates/blog.tsx]:
```suggestion
  const { edges: posts } = data.allGhostPost
```

File: src/templates/blog.tsx [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+
+  return (
+    <Layout>
+      <SEO title="Blog" />
+      <Container>
+        <Section>
+          <H1>Blog</H1>
+          <ul>
+            {posts.map(({ node: post }) => (
+              <li key={post.slug}><Link to={`/blog/post/${post.slug}/`}>{post.title}</Link></li>
```
Comment (cacabo) [inline src/templates/blog.tsx]: Can we abstract the route away in the `constants/routes.ts` file?

File: src/templates/post.css [hunk 15]
```diff
@@ -0,0 +1,251 @@
+/*
+ * Much of this CSS comes from github.com/TryGhost/Casper, which is licensed under the MIT License.
```
Comment (cacabo) [inline src/templates/post.css]: Would expand on this comment to explain what the styles are used for

File: src/templates/post.css [hunk 16]
```diff
@@ -0,0 +1,251 @@
...
+}
+
+.kg-width-full figcaption {
+  padding: 0 1.5em;
+}
+
+.post-content {
+  display: flex;
+  flex-direction: column;
+}
```
Comment (cacabo) [inline src/templates/post.css]:
```suggestion
}

```

File: src/templates/post.tsx [hunk 17]
```diff
@@ -0,0 +1,102 @@
+import React from 'react'
+
```
Comment (cacabo) [inline src/templates/post.tsx]:
```suggestion
```

File: src/templates/post.tsx [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+import { graphql } from 'gatsby'
+import styled from 'styled-components'
+
```
Comment (cacabo) [inline src/templates/post.tsx]: Filename should be capitalized for tsx / component-containing files

File: src/templates/post.tsx [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+            <div
+              className="post-content"
+              dangerouslySetInnerHTML={{
+                __html: `
+${post.codeinjection_head || ''}
+${post.html}
+${post.codeinjection_foot || ''}
+                `,
+              }}
+            />
```
Comment (cacabo) [inline src/templates/post.tsx]: indentation seems off here—I'd put computing the HTML above the render function or in it's own helper function

File: src/templates/post.tsx [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+
+  return (
+    <Layout>
+      <SEO title={post.title} />
+      <MediumContainer>
+        <Section>
+          <H1>{post.title}</H1>
+          <pre>{JSON.stringify(authors)}</pre>
+          <img src={post.feature_image} />
+          <div className={'post-full-content content'}>
```
Comment (cacabo) [inline src/templates/post.tsx]: Destructure post.title and everything else you use before rendering, would also be good to use a typescript interface for this which matches the structure of the graphql query so it's easy to see what fields are available to us in VSCode

File: src/templates/tag.tsx [hunk 21]
```diff
@@ -0,0 +1,80 @@
+import React from 'react'
+import { graphql, Link } from 'gatsby'
+import styled from 'styled-components'
+
+import Layout from '../components/Layout'
+import SEO from '../components/SEO'
+import { IMember, Subset } from '../shared/Icons/types'
+import {
+  Section,
+  H1,
```
Comment (cacabo) [inline src/templates/tag.tsx]: Also capitalize this file name pls :)

File: src/templates/tag.tsx [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+} from '../constants/measurements'
+import { DARK_GRAY } from '../constants/colors'
+
+const TagTemplate = ({ data, location, pageContext }) => {
+  const tag = data.ghostTag
+  const posts = data.allGhostPost.edges
+
+  return (
+    <Layout>
+      <SEO title={tag.name} />
```
Comment (cacabo) [inline src/templates/tag.tsx]: Would be good to destructure name and such above the render method then just use it directly (similar to comment made on other file above)

File: src/templates/tag.tsx [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+      <MediumContainer>
+        <Section>
+          <h1>{tag.name}</h1>
+          <ul>
+            {posts.map(({ node: post }) => (
+              <li key={post.slug}><Link to={`/blog/post/${post.slug}/`}>{post.title}</Link></li>
```
Comment (cacabo) [inline src/templates/tag.tsx]: Abstract away routing to `constants/routes.ts`

**Review by cacabo — CHANGES_REQUESTED** [review]
Insane progress, super excited to polish this up then yeet it

File: gatsby-node.js (hunk 6 above)
Comment (davish) [reply gatsby-node.js]: A lot of this code was copy/pasted from the gatspy-ghost starter repo... this code for doing pagination was just completely unnecessary, since they also imported a library for gatsby pagination. It's been removed/replaced with the library code.

File: gatsby-node.js (hunk 9 above)
Comment (davish) [reply gatsby-node.js]: you're right -- again left over copy/pasted cruft

File: gatsby-config.js [hunk 24]
```diff
@@ -39,6 +39,14 @@ module.exports = {
         },
       },
     },
+    {
+      resolve: `gatsby-source-ghost`,
+      options: {
+        apiUrl: `https://ghost.pennlabs.org`,
+        contentApiKey: process.env.GHOST_API_KEY,
```
Comment (cacabo) [inline gatsby-config.js:54]: Is this in the .env.production and .env.development files? https://www.gatsbyjs.org/docs/environment-variables/#client-side-javascript

File: gatsby-node.js [hunk 25]
```diff
@@ -11,10 +11,17 @@ const path = require(`path`)
 const crypto = require('crypto')
 const remark = require('remark')
 const html = require('remark-html')
+const { paginate } = require('gatsby-awesome-pagination')
+const { postsPerPage } = require('./src/constants/blog.ts')
 
 const markdownProcessor = remark().use(html)
```
Comment (cacabo) [inline gatsby-node.js]: This should come after all imports. there should be no newlines between all relative imports and can be a new line between package imports and relative imports

File: gatsby-node.js [hunk 26]
```diff
@@ -137,20 +201,74 @@ exports.createPages = async ({ graphql, actions }) => {
...
+            slug
+            url
+            postCount
+          }
+        }
+      }
+    }
+  `)
+
+  // Check for any errors
```
Comment (cacabo) [inline gatsby-node.js]: Maybe
```js
const { error } = results
...
```

File: src/constants/blog.ts [hunk 27]
```diff
@@ -0,0 +1 @@
+exports.postsPerPage = 10
```
Comment (cacabo) [inline src/constants/blog.ts]:
```suggestion
exports.postsPerPage = 10

```

Comment (cacabo) [reply src/constants/blog.ts]: Also can't this be `export const postsPerPage = 10`?

File: src/templates/BlogIndex.tsx [hunk 28]
```diff
@@ -0,0 +1,46 @@
...
+  const { edges: posts} = data.allGhostPost
+
+  return (
+    <Layout>
+      <SEO title="Blog" />
+      <Container>
+        <Section>
+          <H1>Blog</H1>
+          <ul>
+            {posts.map(({ node: post }) => (
```
Comment (cacabo) [inline src/templates/BlogIndex.tsx]:
```suggestion
            {posts.map(({ node: { slug, title }}) => (
```

File: src/templates/BlogIndex.tsx [hunk 29]
```diff
@@ -0,0 +1,46 @@
...
+              <li key={post.slug}><Link to={BLOG_POST_ROUTE(post.slug)}>{post.title}</Link></li>
```
Comment (cacabo) [inline src/templates/BlogIndex.tsx]:
```suggestion
              <li key={slug}><Link to={BLOG_POST_ROUTE(slug)}>{title}</Link></li>
```

File: src/templates/Member.tsx [hunk 30]
```diff
@@ -149,13 +161,17 @@ const MemberTemplate = ({ data }: IMemberTemplateProps) => {
       website,
       year_joined: yearJoined,
     },
+    allGhostPost: {
+      edges: postEdges
+    }
   } = data
 
   const roleNames = roles.map(({ name: roleName }) => roleName)
+  const posts = postEdges.map(({ node }) => ({ ...node }))
```
Comment (cacabo) [inline src/templates/Member.tsx]: Why not just
```suggestion
  const posts = postEdges.map(({ node }) => node)
```

File: src/templates/Member.tsx [hunk 31]
```diff
@@ -187,13 +203,26 @@ const MemberTemplate = ({ data }: IMemberTemplateProps) => {
...
           </Row>
         </Section>
+        {posts.length > 0 ? (
+          <Section>
+            <H2>
+              Writings
+            </H2>
+            <P>
+              <ul>
+                {posts.map(post => (<li key={post.slug}><Link to={BLOG_POST_ROUTE(post.slug)}>{post.title}</Link></li>))}
```
Comment (cacabo) [inline src/templates/Member.tsx]:
```suggestion
                {posts.map(({ slug, title }) => (<li key={slug}><Link to={BLOG_POST_ROUTE(slug)}>{title}</Link></li>))}
```

**Review by cacabo — CHANGES_REQUESTED** [review]
Making good progress

Can you also include some screenshots in the PR when you are ready?

File: src/constants/blog.ts (hunk 27 above)
Comment (davish) [reply src/constants/blog.ts]: I was getting an error with that, since we're importing it to `gatsby-node.js`. Not sure if it's related to webpack or the version of Node I have installed locally. I'll try updating node and changing this

File: gatsby-config.js (hunk 24 above)
Comment (davish) [reply gatsby-config.js:54]: Since the key is only used during build time, it doesn't seem like it needs to be included in the client side files.

Comment (cacabo) [reply gatsby-config.js:54]: yep makes sense

File: src/constants/blog.ts (hunk 27 above)
Comment (cacabo) [reply src/constants/blog.ts]: Eek yeah that makes sense. I guess we have to live with that. Annoying that gatsby-node doesn't have TS support. Maybe just put a comment at the top of the file explaining why we are using that syntax (so that it can be used both at build and runtime) and then we should be gtg.

File: gatsby-node.js [hunk 32]
```diff
@@ -116,13 +124,70 @@ exports.sourceNodes = async ({ actions }) => {
...
+        name: `member___NODE`,
+        value: `Labs__Member__${node.slug}`,
+      })
+    }
+  }
+}
+
+const createTagPages = (tags, createPage) => {
+  tags.forEach(({ node }) => {
+    const totalPosts = node.postCount !== null ? node.postCount : 0
```
Comment (cacabo) [inline gatsby-node.js]:
```suggestion
    const totalPosts = node.postCount || 0
```

File: gatsby-node.js [hunk 33]
```diff
@@ -116,13 +124,70 @@ exports.sourceNodes = async ({ actions }) => {
...
+      },
+      context: {
+        slug: node.slug,
+      },
+    })
+  })
+}
+
+const createPostPages = (posts, createPage) => {
+  posts.forEach(({ node }) => {
```
Comment (cacabo) [inline gatsby-node.js]:
```suggestion
  posts.forEach(({ node: { url, slug, authors } }) => {
```

File: gatsby-node.js [hunk 34]
```diff
@@ -142,27 +207,34 @@ exports.createPages = async ({ graphql, actions, reporter }) => {
...
-  await ids.map(id =>
+  const ids = edges.map(({ node: { id, url } }) => ({ id, url }))
+  await ids.map(({ id, url }) =>
     createPage({
-      path: `/team/${id}`,
+      path: `/team/${url}`,
       component: MemberTemplate,
       context: {
         // Data passed to context is available in page queries as GraphQL vars
         id: id,
```
Comment (cacabo) [inline gatsby-node.js]:
```suggestion
        id,
```

File: gatsby-node.js [hunk 35]
```diff
@@ -142,27 +207,34 @@ exports.createPages = async ({ graphql, actions, reporter }) => {
...
+        url: url,
```
Comment (cacabo) [inline gatsby-node.js]:
```suggestion
        url,
```

File: src/components/Blog/Byline.tsx [hunk 36]
```diff
@@ -0,0 +1,75 @@
...
+  // an event in gatsby-node.js.
+  const slugToMember = {}
+  members.forEach(({ node: m }) => {
+    slugToMember[m.url] = m
+  })
+
+  if (!authorsAsMembers) {
+    authorsAsMembers = authors
+      .map(({ slug }) => slugToMember[slug])
+      .filter(mem => !!mem)
```
Comment (cacabo) [inline src/components/Blog/Byline.tsx]:
```suggestion
      .filter(mem => Boolean(mem))
```

File: src/components/Blog/Byline.tsx [hunk 37]
```diff
@@ -0,0 +1,75 @@
...
+
+  if (authorsAsMembers.length === 0) {
+    return <BylineContainer />
+  }
+
+  return (
+    <BylineContainer>
+      By{' '}
+      {authorsAsMembers.map(({ url, student: { name } }) => (
+        <InlineListElement>
```
Comment (cacabo) [inline src/components/Blog/Byline.tsx]: This should have a `key` prop

File: src/components/Blog/MemberBio.tsx [hunk 38]
```diff
@@ -0,0 +1,68 @@
...
+    photo = '',
+    bio = '',
+    student: { name },
+  },
+}: IMemberBioProps): React.ReactElement => (
+  <Card bordered>
+    <CenteredFlex>
+      <div>
+        <P>{name}</P>
+        <P sm mb1 dangerouslySetInnerHTML={{ __html: bio }} />
```
Comment (cacabo) [inline src/components/Blog/MemberBio.tsx]: I think this should be a `div`—in the DOM you technically should not have `p` tags within `p` tags, for example

File: src/shared/Tag.tsx [hunk 39]
```diff
@@ -29,3 +30,19 @@ export const Tags = ({ tags }: ITagsProps) => {
...
+
+interface ILinkedTagsProps {
+  tagToUrl: object
+}
+export const LinkedTags = ({ tagToUrl }: ILinkedTagsProps) => {
+  if (!tagToUrl) return null
+  return (
+    <span style={{ marginBottom: '-4rem' }}>
+      {Object.keys(tagToUrl).map((tagText, i) => (
+        <Link to={tagToUrl[tagText]}>
```
Comment (cacabo) [inline src/shared/Tag.tsx]: I think `key` should be on the `Link`

File: src/templates/Post.tsx [hunk 40]
```diff
@@ -0,0 +1,114 @@
...
+    title,
+    feature_image,
+    html,
+    codeinjection_head,
+    codeinjection_foot,
+    excerpt,
+    tags,
+  } = data.ghostPost
+
+  const authors = data.allMember.nodes
```
Comment (cacabo) [inline src/templates/Post.tsx]:
```suggestion
  const { nodes: authors } = data.allMember
```

**Review by cacabo — CHANGES_REQUESTED** [review]
Pretty much there, just a few small things and we are clear to yeet
